src/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs

```diff
@@ -29,66 +29,75 @@ public TestService(string serviceName)
             // We cannot easily test these so disable the events
             this.CanHandleSessionChangeEvent = false;
             this.CanHandlePowerEvent = false;
-        }
 
-        public static string GetLogPath(string serviceName)
-        {
-            return typeof(TestService).Assembly.Location + "." + serviceName + ".log";
+            this.Server = new NamedPipeServerStream(serviceName);
+            this.Server.WaitForConnectionAsync();
         }
 
+        public NamedPipeServerStream Server { get; set; }
+
         protected override void OnContinue()
         {
-            WriteLog(nameof(OnContinue));
             base.OnContinue();
+            WriteStream(PipeMessageByteCode.Continue);
         }
 
         protected override void OnCustomCommand(int command)
         {
-            WriteLog(nameof(OnCustomCommand) + " command=" + command);
             base.OnCustomCommand(command);
+            WriteStream(PipeMessageByteCode.OnCustomCommand, command);
         }
 
         protected override void OnPause()
         {
-            WriteLog(nameof(OnPause));
             base.OnPause();
+            WriteStream(PipeMessageByteCode.Pause);
         }
 
         protected override void OnSessionChange(SessionChangeDescription changeDescription)
         {
-            WriteLog(nameof(OnSessionChange) + " change=" + changeDescription.ToString());
             base.OnSessionChange(changeDescription);
         }
 
         protected override bool OnPowerEvent(PowerBroadcastStatus powerStatus)
         {
-            WriteLog(nameof(OnPowerEvent) + " status=" + powerStatus.ToString());
             return base.OnPowerEvent(powerStatus);
         }
 
         protected override void OnShutdown()
         {
-            WriteLog(nameof(OnShutdown));
             base.OnShutdown();
         }
 
         protected override void OnStart(string[] args)
         {
-            File.Delete(GetLogPath(ServiceName));
-
-            WriteLog(nameof(OnStart) + " args=" + string.Join(",", args));
             base.OnStart(args);
+            if (args.Length == 3)
+            {
+                Server.WaitForConnectionAsync();
+                WriteStream(PipeMessageByteCode.Start);
+            }
         }
 
         protected override void OnStop()
         {
-            WriteLog(nameof(OnStop));
             base.OnStop();
+            WriteStream(PipeMessageByteCode.Stop);
         }
 
-        private void WriteLog(string msg)
+        private void WriteStream(PipeMessageByteCode code, int command = 0)
         {
-             File.AppendAllText(GetLogPath(ServiceName), msg + Environment.NewLine);
+            if (code == PipeMessageByteCode.OnCustomCommand)
+            {
+                Server.WriteByte((byte)command);
+            }
+            else
+            {
+                byte data = (byte)code;
```

Comment:
Tiny nit, in the one above you have the cast inline, in this one you do not. Consistency is good..